[14.0] Shopfloor: add app backend with automatic routing registration

simahawk opened the pull request:
Depends on
- [x] https://github.com/OCA/wms/pull/315 (Included here)
- [x] https://github.com/OCA/rest-framework/pull/214
- [x] https://github.com/OCA/rest-framework/pull/208
- [x] https://github.com/OCA/edi/pull/469
- [x] https://github.com/OCA/rest-framework/pull/219

The main change that this PR brings is: you can now have multiple shopfloor apps and configure them differently (eg: different url, different menus, different auth, etc).

NOTE: if you want to support the old URL of the app for your own project, you'll have to add your own controller, like this:

```
from odoo import http
from odoo.addons.shopfloor_mobile_base.controllers.main import (
    ShopfloorMobileAppMixin,
)


class ShopfloorMobileAppController(http.Controller, ShopfloorMobileAppMixin):
    @http.route(
        ["/shopfloor_mobile/app", "/shopfloor_mobile/app/<string:demo>"],
        auth="public",
    )
    def load_app(self, demo=False, **kw):
        # keep the old URL for backward compat
        # after v14 refactoring w/ shopfloor.app
        shopfloor_app = http.request.env.ref("my.new.shopfloor.app")
        return self._load_app(
            shopfloor_app, demo=True if demo else False, **kw
        )

```
I've chosen not to support this by default because the app can be created in several ways and there's no default app anymore.

simahawk committed:
shopfloor_base: add app backend

This is the base to provide specific settings by app.
For instance, today you cannot control which users
can use the app or provide any extra global settings
without adding a specific parameters.

Opens the door to better management of the app
and to manage multiple apps on the same instance.

simahawk committed:
sf_mobile_base: fix _get_app_version in dev mode

When developing the path of the module might be not the right one
and you might get an error and fallback to 'dev' version.

simahawk committed:
shopfloor_base: test files better naming

simahawk committed:
shopfloor_base: get rid of static controller

Use endpoint_route_handler capabilities to register endpoints on the fly.

simahawk committed:
sf_mobile_base: adapt to shopfloor_base changes

simahawk committed:
shopfloor*_api_key: limit access to specific keys

simahawk committed:
shopfloor.app: rework route registration

simahawk committed:
shopfloor.app mobile: rework route registration

simahawk committed:
shopfloor.app: index tech_name

simahawk committed:
shopfloor_base: add default app

simahawk committed:
sf_base: shopfloor.app load services when available

simahawk committed:
sf_base: tests unify service get

simahawk committed:
sf_base: shopfloor.app fix and test endpoint registration

simahawk committed:
sf_base: shopfloor.app display registered routes

simahawk committed:
sf_mobile_base: shopfloor.app required now

simahawk committed:
shopfloor: fix menu service override

simahawk committed:
shopfloor: unify service lookup in tests

simahawk committed:
shopfloor.app: register only static routes

Static routes are way better as we don't risk to mess up
identical route paths w/ different kinds of auth.
It's also easier to handle the cleanup of the routes
when we have to.

This makes having 2 api_route fields useless
and allows to show real paths in the developer tab
which is way better in terms of end user info.

simahawk committed:
sf_mobile_auth_api_key: ease access to groups

simahawk committed:
sf_base: limit profiles and menu items by app

simahawk committed:
shopfloor: adapt to sf_base, add more demo data

simahawk committed:
sf_mobile_base: profile required only when needed

The shopfloor.app now states if the profile is required.
If not, don't bother asking for a profile.

simahawk committed:
shopfloor_base: add per app api-docs

simahawk committed:
shopfloor_base: apps menu 1st

simahawk committed:
shopfloor_base: fix test service setup

simahawk committed:
shopfloor.app: fix manifest url

simahawk committed:
shopfloor: remove old mig files for 13.0

simahawk committed:
shopfloor: don't depend on api key

simahawk committed:
shopfloor_base: get rid of api key left overs

simahawk committed:
shopflor.app: add category

simahawk committed:
shopfloor: add default app migration

simahawk committed:
sf_mobile_base_auth_api_key: add specific demo data

simahawk committed:
sf_mobile_base_auth_user: add specific demo data

simahawk committed:
shopfloor_base: 14.0.2.0.0

simahawk committed:
shopfloor.app: add btn to view menu items

simahawk committed:
shopfloor.app: improve form view

simahawk committed:
sf_mobile_base: get rid of backward compat controller

simahawk committed:
shopfloor_mobile_base: 14.0.2.0.0

simahawk committed:
shopfloor_example: add example app

simahawk committed:
shopfloor_ui_tests: FIXME

jbaudoux commented on shopfloor_base/models/shopfloor_app.py:
```diff
@@ -0,0 +1,262 @@
+    url = fields.Char(compute="_compute_url", help="Public URL to use the app.")
+    api_docs_url = fields.Char(compute="_compute_url", help="Public URL for api docs.")
+    auth_type = fields.Selection(
+        selection="_selection_auth_type", default="user_endpoint"
```
The default value needs the module shopfloor_mobile_base_auth_user that is not a dependency

simahawk replied:
no, this comes from `endpoint_route_handler` the way you implement it in the front is up to you.

jbaudoux commented on shopfloor_base/models/shopfloor_app.py:
```diff
@@ -0,0 +1,262 @@
+            rec.profile_required = bool(rec.profile_ids)
+
+    def _selection_auth_type(self):
+        return self.env["endpoint.route.handler"]._selection_auth_type()
```
user_endoint needs the module shopfloor_mobile_base_auth_user
Either it should then be in the dependencies or that selection option should be added by that module.

jbaudoux replied:
Even with the module installed, I still get:
[Vue warn]: Error in created hook: "Error: Auth type 'user_endpoint' not supported"

simahawk replied:
I think this is solved now.

simahawk committed:
shopfloor: 14.0.2.0.0

simahawk committed:
shopfloor.app: comp registry check reusable

simahawk committed:
sf_mobile_base: fix demo mode OdooMocked

simahawk committed:
shopfloor: fix demo data usage in tests

simahawk committed:
shopfloor_base: fix common setup

simahawk committed:
shopfloor: override shopfloor.app in tests

simahawk committed:
sf_delivery_shipment: adapt to shpofloor.app setup

simahawk committed:
sf_manual_product_transfer: adapt to shpofloor.app setup

simahawk committed:
sf_workstation: adapt to shpofloor.app setup

simahawk committed:
sf_batch_automatic_creation: adapt to shpofloor.app setup

sebalix commented:
@simahawk I pushed your patch in https://github.com/OCA/wms/pull/316/commits/aa64e673e4318324a9b709a8faac5ecb38d0ad79 (with a fix in the query parameter to make it a tuple)

simahawk committed:
shopfloor: fix upgrade to 2.0.0

simahawk committed:
shopfloor_batch_automatic_creation 14.0.1.1.0

simahawk committed:
shopfloor_manual_product_transfer 14.0.1.2.0

simahawk committed:
shopfloor_workstation: dev status -> Beta

simahawk committed:
shopfloor_workstation 14.0.1.2.0

sebalix reviewed (approved):
Tested some Shopfloor scenario locally :+1:

sebalix commented on shopfloor_base/models/shopfloor_app.py:
```diff
@@ -0,0 +1,265 @@
+        }
+
+    def action_view_menu_items(self):
+        action = self.env.ref("shopfloor_base.action_shopfloor_menu").read()[0]
```
```suggestion
        action = self.env["ir.actions.act_window"]._for_xml_id("shopfloor_base.action_shopfloor_menu")
```

sebalix commented on shopfloor_mobile_base_auth_api_key/models/shopfloor_app.py:
```diff
@@ -0,0 +1,30 @@
+        return self.auth_api_key_group_ids.auth_api_key_ids.ids
+
+    def action_manage_api_key_groups(self):
+        action = self.env.ref(
+            "auth_api_key_group.auth_api_key_group_act_window"
+        ).read()[0]
```
```suggestion
        action = self.env["ir.actions.act_window"]._for_xml_id("auth_api_key_group.auth_api_key_group_act_window")
```

simahawk committed:
sf_mobile_base: use sf.app name

simahawk committed:
sf_mobile_base: fix demo mode enable

simahawk committed:
sf_mobile_base: use DotDict to ease access to info

simahawk committed:
sf_mobile_base: fix html title

simahawk committed:
sf_mobile_base: get version from sf.app

simahawk committed:
sf_mobile_base: decouple event_hub

Use event_hub explicitly where needed.

simahawk committed:
sf_mobile_base: emit event on app mounted

simahawk committed:
sf_mobile_base: fetch bypass load state optionally

Sometimes you want to make XHR calls in background
w/out fading out the screen and showing a spinner.
Just making this possible.

simahawk committed:
sf_mobile_base: allow custom loading msg

simahawk committed:
sf_mobile_base: emit sync event

Also, plan renaming /user_config to /sync
as this endpoint can be considered a sync hook
for any kind of data.

simahawk committed:
sf_mobile_base: fix screen bg class

Using grey or other literal colors in :class directive does not work
because Vue tries to transform it and requires hex color there :/

simahawk committed:
sf_mobile_base: improve odoo demo data handler

Prefix all mocked methods to avoid overriding inherited ones (eg: 'get')

simahawk committed:
shopfloor_rest_log: adapt to sf.app

simahawk committed:
shopfloor_rest_log 14.0.1.1.0

simahawk committed:
shopfloor_rest_log: add action on app view

simahawk committed:
sf_mobile_base: make form mixin func name homogeneus

simahawk committed:
shopfloor_base: fix form mixin validators

simahawk committed:
shopfloor: drop form_mixin left over

Probably left there by error when I split sf_base out.

sebalix commented:
Rebased to resolve conflicts on `shopfloor` and `shopfloor_base` manifest files

simahawk committed:
shopfloor_mobile_base_auth_api_key: dev status = Beta

simahawk committed:
shopfloor_mobile_base_auth_user: dev status = Beta

simahawk committed:
s_mobile_base: pass root object to login events

simahawk committed:
shopfloor_mobile_base_auth_api_key: 14.0.1.1.0

simahawk committed:
shopfloor_mobile_base_auth_user: 14.0.1.1.0

sebalix reviewed (approved):
But Travis is still unhappy regarding tests...

jbaudoux commented on shopfloor_base/models/shopfloor_app.py:
```diff
@@ -0,0 +1,268 @@
+    url = fields.Char(compute="_compute_url", help="Public URL to use the app.")
+    api_docs_url = fields.Char(compute="_compute_url", help="Public URL for api docs.")
+    auth_type = fields.Selection(
+        selection="_selection_auth_type", default="user_endpoint"
```
Sad you cannot enable multiple auth types

simahawk commented:
I've restored the commits that got lost in the last few force pushes. That's why it was failing.

simahawk commented:
/ocabot merge nobump
